### PR TITLE
Speedy: shortcup failure in SValue ordering.

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
Before this PR, comparison of lists, maps, ans structural record had
always a  linear complexity (even in case of failure).

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [X] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [X] Include appropriate tests
- [X] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 1]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+    @inline
+    def pop(): Unit = {
+      stackX = stackX.tail
+      stackY = stackY.tail
+    }
+
+    @inline
+    def step(tuple: (SValue, SValue)) =
+      tuple match {
+        case (SUnit, SUnit) =>
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:42]: I know this is not necessary but it makes the code easier to read for people with limited Scala experience, of which we have many.
```suggestion
        case (SUnit, SUnit) =>
          ()
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 2]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+        case (SNumeric(x), SNumeric(y)) =>
+          diff = x compareTo y
+        case (SText(x), SText(y)) =>
+          diff = Utf8.Ordering.compare(x, y)
+        case (SDate(x), SDate(y)) =>
+          diff = x compareTo y
+        case (STimestamp(x), STimestamp(y)) =>
+          diff = x compareTo y
+        case (SParty(x), SParty(y)) =>
+          // party are ASCII, so UTF16 comparison matches UTF8 comparison.
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:56]:
```suggestion
          // parties are ASCII, so UTF16 comparison matches UTF8 comparison.
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 3]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+          diff = x compareTo y
+        case (SContractId(x), SContractId(y)) =>
+          diff = compareCid(x, y)
+        case (SEnum(_, _, xRank), SEnum(_, _, yRank)) =>
+          diff = xRank compareTo yRank
+        case (SRecord(_, _, xs), SRecord(_, _, ys)) =>
+          push(xs.iterator().asScala, ys.iterator().asScala)
+        case (SVariant(_, _, xRank, x), SVariant(_, _, yRank, y)) =>
+          push(Iterator.single(x), Iterator.single(y))
+          diff = xRank compareTo yRank
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:66]: Just to reflect the actual order of comparison in the code.
```suggestion
          diff = xRank compareTo yRank
          push(Iterator.single(x), Iterator.single(y))
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 4]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+        case (SList(xs), SList(ys)) =>
+          push(xs.iterator, ys.iterator)
+        case (SOptional(xOpt), SOptional(yOpt)) =>
+          push(xOpt.iterator, yOpt.iterator)
+        case (STextMap(xMap), STextMap(yMap)) =>
+          val xKeys = xMap.keys.toSeq.sorted
+          val yKeys = yMap.keys.toSeq.sorted
+          push(
+            new InterlacedIterator(xKeys.iterator.map(SText), xKeys.iterator.map(xMap)),
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:75]: I assume `STextMap` is backed by an ordered tree map. Is that correct? If so, this gives us complexity `O(n log n)`. Is there a way to get an iterator over the key/value pairs in the order of the keys? This should leave us with complexity `O(n)` then.

Same for generic map below.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 5]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+        case (SGenMap(xMap), SGenMap(yMap)) =>
+          push(
+            new InterlacedIterator(xMap.keys.iterator, xMap.values.iterator),
+            new InterlacedIterator(yMap.keys.iterator, yMap.values.iterator),
+          )
+        case (SStruct(_, xs), SStruct(_, ys)) =>
+          push(xs.iterator().asScala, ys.iterator().asScala)
+        case (SAny(xType, x), SAny(yType, y)) =>
+          push(Iterator.single(x), Iterator.single(y))
+          diff = TypeOrdering.compare(xType, yType)
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:87]: Just to reflect the actual order of comparison in code.
```suggestion
          diff = TypeOrdering.compare(xType, yType)
          push(Iterator.single(x), Iterator.single(y))
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala [hunk 6]
```diff
@@ -5,22 +5,105 @@ package com.daml.lf
...
+        case (STypeRep(xType), STypeRep(yType)) =>
+          diff = TypeOrdering.compare(xType, yType)
+        case _ =>
+          throw SError.SErrorCrash("functions are not comparable")
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:91]: I would prefer something along the lines of the code below in order to separate behaviour that users can trigger from buggy behaviour of the interpreter.
```suggestion
        case (_: SPAP, _: SPAP) =>
          throw SError.SErrorCrash("functions are not comparable")
        // We should never hit this case at runtime.
        case _ =>
          throw SError.SErrorCrash("BUG: comparison of incomparable values")
```

**Review by hurryabit — APPROVED** [review]
Nice. Thank you.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala (hunk 4 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:75]: Inside Speedy, `STextMap` is based on HashMap, so there is no way to get linear complexity here.
Note that #7334, replaces  `STextMap` by `SGenMap` (based on TreeMap) so we will get linear complexity here.

Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala:76]: That's good to know. Thanks a lot for the explanation.

**Pull request merged**